crd-generator/collector/src/main/java/io/fabric8/crd/generator/collector/CustomResourceCollector.java

```diff
@@ -0,0 +1,158 @@
+    if (classNames != null) {
+      classNames.stream()
+          .filter(Objects::nonNull)
+          .forEach(this.customResourceClassNames::add);
+    }
+    return this;
+  }
+
+  public CustomResourceCollector withIndex(IndexView... index) {
+    jandexCustomResourceClassScanner.withIndex(index);
+    return this;
+  }
+
+  public CustomResourceCollector withIndices(Collection<IndexView> indices) {
+    jandexCustomResourceClassScanner.withIndices(indices);
+    return this;
+  }
+
+  public CustomResourceCollector withFileToIndex(File... files) {
+    jandexCustomResourceClassScanner.withFileToIndex(files);
+    return this;
+  }
+
+  public CustomResourceCollector withFilesToIndex(Collection<File> files) {
+    jandexCustomResourceClassScanner.withFilesToIndex(files);
+    return this;
+  }
+
+  public CustomResourceCollector withForceIndex(boolean forceIndex) {
+    jandexCustomResourceClassScanner.withForceIndex(forceIndex);
+    return this;
+  }
+
+  public CustomResourceCollector withIncludePackages(Collection<String> packages) {
+    if (packages != null) {
+      packages.stream()
+          .filter(Objects::nonNull)
+          .map(pkg -> (Predicate<String>) s -> s.startsWith(pkg))
+          .reduce(Predicate::or)
+          .ifPresent(this.classNamePredicates::add);
+    }
+    return this;
+  }
+
+  public CustomResourceCollector withExcludePackages(Collection<String> packages) {
+    if (packages != null) {
+      packages.stream()
+          .filter(Objects::nonNull)
+          .map(pkg -> (Predicate<String>) s -> !s.startsWith(pkg))
+          .reduce(Predicate::or)
+          .ifPresent(this.classNamePredicates::add);
+    }
+    return this;
+  }
+
+  private List<Class<? extends HasMetadata>> findCustomResourceClassesAsList() {
+    Set<String> customResourcesClassNames = new HashSet<>(customResourceClassNames);
+
+    // search only if custom resource class names are not explicitly given
+    if (customResourcesClassNames.isEmpty()) {
+      customResourcesClassNames.addAll(jandexCustomResourceClassScanner.findCustomResourceClasses());
+      log.debug("Found {} custom resource classes", customResourcesClassNames.size());
+    } else {
+      log.debug("Using explicit {} custom resource classes and skip scanning", customResourcesClassNames);
+    }
+
+    Predicate<String> classNamePredicate = classNamePredicates.stream()
+        .reduce(Predicate::and)
+        .orElse(className -> true);
+
+    List<Class<? extends HasMetadata>> customResourceClasses = customResourcesClassNames.stream()
+        .filter(classNamePredicate)
+        .map(customResourceClassLoader::loadCustomResourceClass)
+        .collect(Collectors.toList());
+
+    log.debug("Found {} custom resource classes after filtering", customResourceClasses.size());
+    return customResourceClasses;
+  }
+
+  public Class<? extends HasMetadata>[] findCustomResourceClasses() {
```

Comment:
This method is required because the CRDGenerator accepts only an array.

Can we extend the CRDGenerator with an additional method?

```java
  public CRDGenerator customResourceClasses(Collection<Class<? extends HasMetadata>> crClasses) {
    return customResources(crClasses.stream().map(CustomResourceInfo::fromClass).toArray(CustomResourceInfo[]::new));
  }
```

Reply:
@baloo42 yes, feel free to add whatever method works best.

Reply:
Note that we could possibly make API changes in the v2 version if needed… as long as it is possible to generate CRDs programmatically, incrementally registering classes to generate from.

Reply:
I added two additional input methods for collections (both nullsafe) and kept the existing ones, so that upstream projects can now choose between a collection, an array or a single CustomResourceInfo or CustomResource class.

> Note that we could possibly make API changes in the v2 version if needed… as long as it is possible to generate CRDs programmatically, incrementally registering classes to generate from.

Good to know, but IMHO we should keep both possibilities for now, so that we don't introduce unecessary changes. Please let me now if I should mark the array methods as deprecated, otherwise this thread can be marked as resolved in my opinion.